### PR TITLE
Add the option to download tasks from the cli

### PR DESCRIPTION
**Pull request by AugustoPeres**
Adds the option to download tasks from the cli.

Usage examples:

* `inductiva tasks download id_1 id_2 --filenames stderr.txt stdout.txt`
* `inductiva tasks download id`
* `inductiva tasks download id --output_dir some_output_dir`
* `inductiva tasks download id_1 id_2 --output_dir some_output_dir` -> This will fail

### CODE REVIEW

**Review by hpenedones — COMMENTED** [review]
This is super useful! Thanks Augusto.

**Review by Pbarbosa92 — APPROVED** [review]
LGTM!

File: inductiva/_cli/cmd_tasks/download.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+"""Downloads a tasks by id via CLI."""
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:1]:
```suggestion
"""Download the outputs of a task by ID via CLI."""
```

File: inductiva/_cli/cmd_tasks/download.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import argparse
+import inductiva
+
+
+def download_task(args):
+    """Downloads a task by id."""
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:7]:
```suggestion
    """Download the outputs of a task by ID."""
```

File: inductiva/_cli/cmd_tasks/download.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+"""Downloads a tasks by id via CLI."""
+import argparse
+import inductiva
+
+
+def download_task(args):
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:6]: Let's keep the same structure of the `tasks` module. Moreover, this is downloading the results of the simulation and not any metadata of the task.
```suggestion
def download_outputs(args):
```

File: inductiva/_cli/cmd_tasks/download.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """Downloads a task by id."""
+    ids = args.id
+    filenames = args.filenames
+    output_dir = args.output_dir
+
+    if not ids:
+        print("No id(s) specified.\n"
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:13]:
```suggestion
        print("No ID(s) specified.\n"
```

File: inductiva/_cli/cmd_tasks/download.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+              "> Use `inductiva tasks download -h` for help.")
+        return 1
+
+    ids = set(ids)
+
+    if len(ids) > 1 and output_dir is not None:
+        print("Cannot specify `output_dir` for more than one download")
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:20]: Why not? I would actually like to just download all of these tasks into my own specified dir, with task_ids folders inside.

File: inductiva/_cli/cmd_tasks/download.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        return 1
+
+    for task_id in ids:
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:23]: Be consistent with the names, if here one uses `task_id` than the set should be `task_ids` or `task_ids_list`.

File: inductiva/_cli/cmd_tasks/download.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        try:
+            inductiva.tasks.Task(task_id).download_outputs(
+                filenames=filenames, output_dir=output_dir)
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:26]: Can you verify what happens if we download a set of tasks where the filename doesn't exist in only one of the tasks? This is also important for the tasks.

File: inductiva/_cli/cmd_tasks/download.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    """Register the download task command."""
+
+    subparser = parser.add_parser("download",
+                                  help="Download tasks.",
+                                  formatter_class=argparse.RawTextHelpFormatter)
+
+    subparser.description = (
+        "The `inductiva tasks download` command downloads specified tasks "
+        "on the platform.\n"
+        "You can download multiple tasks by passive multiple ids.\n")
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:42]:
```suggestion
        "The `inductiva tasks download` command allows to download the outputs of specified tasks "
        "on the platform.\n"
        "You can download multiple tasks by passing multiple ids.\n")
```

**Comment by AugustoPeres** [comment]
## QA

1. Install from this branch;
2. Run the commands in the pull request's description. You should observe that:
   1. By default tasks go to /inductiva_outputs/task_id/;
   2. If a directory is provided the output goes instead to `./give_output_dir/`
   3. It will not allow the user to provide multiple IDs *and* and output dir
   4. When using `--filenames` only those files are downloaded

File: inductiva/_cli/cmd_tasks/download.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                           type=str,
+                           help="ID(s) of the task(s) to download.",
+                           nargs="*")
+    subparser.add_argument("--filenames",
+                           type=str,
+                           help="Names of the files to download.",
+                           nargs="*")
+    subparser.add_argument("--output_dir",
+                           type=str,
+                           help="Where to download the task to.")
```
Comment (IvanPombo) [inline inductiva/_cli/cmd_tasks/download.py:54]:
```suggestion
                           help="Path of where to download the task outputs..")
```

**Review by IvanPombo — COMMENTED** [review]
Simple comments for readibility.

**Comment by Pbarbosa92** [comment]
**QA**
Just one question on the second test. Apart from that works great!
- `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved` ✅ 
- `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved --output_dir my_task_output` ☑️  The output goes into /inductiva_output/my_task_output/<task_id> is this expected ?
- `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved g24wos6vqjc6jhq0jecox8d1m --output_dir my_task_output` ✅ 
- `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved --filenames stderr.txt stdout.txt` ✅ Only downloads the specified files

File: inductiva/_cli/cmd_tasks/download.py (hunk 5 above)
Comment (AugustoPeres) [reply inductiva/_cli/cmd_tasks/download.py:20]: That is a good point. Will add that

File: inductiva/_cli/cmd_tasks/download.py (hunk 7 above)
Comment (AugustoPeres) [reply inductiva/_cli/cmd_tasks/download.py]: It fails for that task and continues the downloads for the others

**Comment by AugustoPeres** [comment]
> **QA** Just one question on the second test. Apart from that works great!
> 
> * `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved` ✅
> * `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved --output_dir my_task_output` ☑️  The output goes into /inductiva_output/my_task_output/<task_id> is this expected ?
> * `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved g24wos6vqjc6jhq0jecox8d1m --output_dir my_task_output` ✅
> * `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved --filenames stderr.txt stdout.txt` ✅ Only downloads the specified files

New QA instructions:
*`inductiva tasks download 6yndq10u2j28px5k5i6hm9ved --output_dir my_task_output` ☑️  The output goes into /inductiva_output/my_task_output/<task_id> is this expected ? Yes
* `inductiva tasks download 6yndq10u2j28px5k5i6hm9ved g24wos6vqjc6jhq0jecox8d1m --output_dir my_task_output`. Now, instead of failing, this goes `inductiva_output/output_dir/task_id`

File: inductiva/_cli/cmd_tasks/download.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    task_ids = set(task_ids)
+
+    for task_id in task_ids:
+        if output_dir is not None:
+            output_dir_task = pathlib.Path(output_dir) / task_id
+        else:
+            output_dir_task = None
+        try:
+            inductiva.tasks.Task(task_id).download_outputs(
+                filenames=filenames, output_dir=output_dir_task)
```
Comment (sergiomsantos) [inline inductiva/_cli/cmd_tasks/download.py:28]:
```suggestion
            task = inductiva.tasks.Task(task_id)
            task.download_outputs(filenames=filenames,
                                  output_dir=output_dir_task)
```

File: inductiva/_cli/cmd_tasks/download.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    subparser = parser.add_parser("download",
+                                  help="Download tasks.",
+                                  formatter_class=argparse.RawTextHelpFormatter)
+
+    subparser.description = (
+        "The `inductiva tasks download` command allows to download the outputs"
+        " of specified tasks "
+        "on the platform.\n"
+        "You can download multiple tasks by passing multiple ids.\n")
```
Comment (sergiomsantos) [inline inductiva/_cli/cmd_tasks/download.py:45]:
```suggestion
        "Download the output files produced in the context of the tasks with the\n"
        "given ID(s). All files are downloaded unless the --filenames list is\n"
        "given, in which case only the indicated files for each task are\n"
        "downloaded. The name of the output folder can be configured through\n"
        "the --output_dir option. Files are downloaded to a subdirectory named\n"
        "after the ID of the corresponding task.")
```

Not sure if line breaks are needed though.

File: inductiva/_cli/cmd_tasks/download.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+              "> Use `inductiva tasks download -h` for help.")
+        return 1
+
+    task_ids = set(task_ids)
+
+    for task_id in task_ids:
+        if output_dir is not None:
+            output_dir_task = pathlib.Path(output_dir) / task_id
+        else:
+            output_dir_task = None
```
Comment (sergiomsantos) [inline inductiva/_cli/cmd_tasks/download.py:25]: what if `output_dir` is empty?

Comment (sergiomsantos) [reply inductiva/_cli/cmd_tasks/download.py:25]: `output_dir` can indeed be empty: `download --output_dir ''`

File: inductiva/_cli/cmd_tasks/download.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                                  formatter_class=argparse.RawTextHelpFormatter)
+
+    subparser.description = (
+        "Download the output files produced in the context of the tasks with "
+        "the given ID(s).\nAll files are downloaded unless the --filenames"
+        " list is given, in which case only\n the indicated files for each task"
+        " are downloaded.\n"
+        "The name of the output folder can be configured "
+        "through the --output_dir option.\nFiles are downloaded to a "
+        "subdirectory named after the ID of the corresponding task.")
```
Comment (sergiomsantos) [inline inductiva/_cli/cmd_tasks/download.py:50]: why do  line breaks have to exist in the middle of each line and not at the end?
Makes reading it harder.

**Review by sergiomsantos — APPROVED** [review]

File: inductiva/_cli/cmd_tasks/download.py (hunk 13 above)
Comment (AugustoPeres) [reply inductiva/_cli/cmd_tasks/download.py:50]: It was to prevent long lines being show in therminal. But I will change that

**Review by sergiomsantos — APPROVED** [review]

**Pull request merged**